Open up AliasAnalysisKind for any ops

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23695 [wip] Fixed size arrays
* #23668 [wip] Call aten ops through c10 dispatcher
* #23888 Allow TensorMethods.h to include Dispatcher.h (alternative)
* #23667 Register ATen ops with c10
* #23748 Disambiguate tensor and string ops
* #23694 toString(FunctionSchema) shows overload name
* #23666 c10 dispatcher stores autograd kernels
* #23665 Allow kernels that don't have a boxed version
* #23447 Unboxed kernels in c10
* **#23834 Open up AliasAnalysisKind for any ops**

Re-expot of reverted PR https://github.com/pytorch/pytorch/pull/22239 (or actually https://github.com/pytorch/pytorch/pull/23810) with the bug fixed

A previous diff removed the special casing for aten:: and prim:: ops in alias analysis and implements alias analysis purely
based on the AliasAnalysisKind. To be sure it doesn't break our existing code base, it added asserts that make sure that
our existing aten:: and prim:: ops set the correct AliasAnalysisKind.

However, we don't need that restriction for future ops. Since we are now certain all existing cases are set up correctly,
we can remove these assertions.

Differential Revision: [D16657239](https://our.internmc.facebook.com/intern/diff/D16657239/)

smessmer committed:
Open up AliasAnalysisKind for any ops

Re-expot of reverted PR https://github.com/pytorch/pytorch/pull/23810 with the bug fixed

A previous diff removed the special casing for aten:: and prim:: ops in alias analysis and implements alias analysis purely
based on the AliasAnalysisKind. To be sure it doesn't break our existing code base, it added asserts that make sure that
our existing aten:: and prim:: ops set the correct AliasAnalysisKind.

However, we don't need that restriction for future ops. Since we are now certain all existing cases are set up correctly,
we can remove these assertions.

Differential Revision: [D16657239](https://our.internmc.facebook.com/intern/diff/D16657239/)

smessmer committed:
Update on "Open up AliasAnalysisKind for any ops"


Re-expot of reverted PR https://github.com/pytorch/pytorch/pull/23810 with the bug fixed

A previous diff removed the special casing for aten:: and prim:: ops in alias analysis and implements alias analysis purely
based on the AliasAnalysisKind. To be sure it doesn't break our existing code base, it added asserts that make sure that
our existing aten:: and prim:: ops set the correct AliasAnalysisKind.

However, we don't need that restriction for future ops. Since we are now certain all existing cases are set up correctly,
we can remove these assertions.

Differential Revision: [D16657239](https://our.internmc.facebook.com/intern/diff/D16657239/)

zdevito reviewed (approved):
(re-approve due to revert)

smessmer committed:
Update on "Open up AliasAnalysisKind for any ops"


Re-expot of reverted PR https://github.com/pytorch/pytorch/pull/22239 (or actually https://github.com/pytorch/pytorch/pull/23810) with the bug fixed

A previous diff removed the special casing for aten:: and prim:: ops in alias analysis and implements alias analysis purely
based on the AliasAnalysisKind. To be sure it doesn't break our existing code base, it added asserts that make sure that
our existing aten:: and prim:: ops set the correct AliasAnalysisKind.

However, we don't need that restriction for future ops. Since we are now certain all existing cases are set up correctly,
we can remove these assertions.

Differential Revision: [D16657239](https://our.internmc.facebook.com/intern/diff/D16657239/)

smessmer committed:
Update on "Open up AliasAnalysisKind for any ops"


Re-expot of reverted PR https://github.com/pytorch/pytorch/pull/22239 (or actually https://github.com/pytorch/pytorch/pull/23810) with the bug fixed

A previous diff removed the special casing for aten:: and prim:: ops in alias analysis and implements alias analysis purely
based on the AliasAnalysisKind. To be sure it doesn't break our existing code base, it added asserts that make sure that
our existing aten:: and prim:: ops set the correct AliasAnalysisKind.

However, we don't need that restriction for future ops. Since we are now certain all existing cases are set up correctly,
we can remove these assertions.

Differential Revision: [D16657239](https://our.internmc.facebook.com/intern/diff/D16657239/)
